### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by duncan4123**

### CODE REVIEW

File: contracts/ExternalBribe.sol [hunk 1]
```diff
@@ -1,63 +1,73 @@
 // SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/ExternalBribe.sol:1]: [Note] changes are only formatting.

File: contractNameAndAddress.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { HardhatRuntimeEnvironment } from 'hardhat/types'
+import { DeployFunction } from 'hardhat-deploy/types'
+
+const func: GetContractNames = async function (hre: HardhatRuntimeEnvironment) {
```
Comment (0xMotto) [inline contractNameAndAddress.ts:4]: [Question] Why is the function name GetContractNames which acutally deploys GaugeFactory?

File: contracts/Flow.sol [hunk 3]
```diff
@@ -42,23 +45,20 @@ contract Velo is IVelo {
...
-    // Initial mint: total 82M
-    //  4M for "Genesis" pools
-    // 30M for liquid team allocation (40M excl init veNFT)
-    // 48M for future partners
+    // NFTs are minted from this amount as well now
     function initialMint(address _recipient) external {
         require(msg.sender == minter && !initialMinted);
         initialMinted = true;
-        _mint(_recipient, 82 * 1e6 * 1e18);
+        _mint(_recipient, 400 * 1e6 * 1e18); //#settings
```
Comment (0xMotto) [inline contracts/Flow.sol:52]: [Note] Initial mint changes from 82M to 400M FLOW token

File: contracts/Flow.sol [hunk 4]
```diff
@@ -1,25 +1,28 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/Flow.sol:1]: [Note] Changes in token name and formatting

File: contracts/FlowGovernor.sol [hunk 5]
```diff
@@ -9,7 +9,7 @@ import {L2GovernorCountingSimple} from "contracts/governance/L2GovernorCountingS
 import {L2GovernorVotes} from "contracts/governance/L2GovernorVotes.sol";
 import {L2GovernorVotesQuorumFraction} from "contracts/governance/L2GovernorVotesQuorumFraction.sol";
 
-contract VeloGovernor is
+contract FlowGovernor is
```
Comment (0xMotto) [inline contracts/FlowGovernor.sol:12]: [Note] only protocol / token name changes

File: contracts/Gauge.sol [hunk 6]
```diff
@@ -1,98 +1,115 @@
...
+    event ClaimFees(address indexed from, uint256 claimed0, uint256 claimed1);
+    event ClaimRewards(
+        address indexed from,
+        address indexed reward,
+        uint256 amount
+    );
+
+    constructor(
+        address _stake,
+        address _internal_bribe,
```
Comment (0xMotto) [inline contracts/Gauge.sol:98]: [Question] Are we going to keep internal bribe as parameter in constructor here?

File: contracts/Gauge.sol [hunk 7]
```diff
@@ -101,35 +118,48 @@ contract Gauge is IGauge {
...
+            ) {
                 fees0 = 0;
                 _safeApprove(_token0, internal_bribe, _fees0);
                 IBribe(internal_bribe).notifyRewardAmount(_token0, _fees0);
             } else {
                 fees0 = _fees0;
             }
-            if (_fees1 > IBribe(internal_bribe).left(_token1) && _fees1 / DURATION > 0) {
+            if (
+                _fees1 > IBribe(internal_bribe).left(_token1) &&
```
Comment (0xMotto) [inline contracts/Gauge.sol:160]: [Question] still internal bribes?

File: contracts/Gauge.sol [hunk 8]
```diff
@@ -101,35 +118,48 @@ contract Gauge is IGauge {
...
+                _fees1 / DURATION > 0
+            ) {
                 fees1 = 0;
                 _safeApprove(_token1, internal_bribe, _fees1);
                 IBribe(internal_bribe).notifyRewardAmount(_token1, _fees1);
```
Comment (0xMotto) [inline contracts/Gauge.sol:165]: [Question] still internal bribes?

File: contracts/Gauge.sol [hunk 9]
```diff
@@ -101,35 +118,48 @@ contract Gauge is IGauge {
...
+            uint256 _fees1 = fees1 + claimed1;
             (address _token0, address _token1) = IPair(stake).tokens();
-            if (_fees0 > IBribe(internal_bribe).left(_token0) && _fees0 / DURATION > 0) {
+            if (
+                _fees0 > IBribe(internal_bribe).left(_token0) &&
+                _fees0 / DURATION > 0
+            ) {
                 fees0 = 0;
                 _safeApprove(_token0, internal_bribe, _fees0);
                 IBribe(internal_bribe).notifyRewardAmount(_token0, _fees0);
```
Comment (0xMotto) [inline contracts/Gauge.sol:155]: [Question] still internal bribes?

File: contracts/Gauge.sol [hunk 10]
```diff
@@ -1,98 +1,115 @@
 // SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/Gauge.sol:1]: [Note] Don't see changes to notify external bribe rewards. Mostly formatting changes.

File: contracts/Minter.sol [hunk 11]
```diff
@@ -1,35 +1,41 @@
...
+    uint256 public active_period;
+    uint256 internal constant LOCK = 86400 * 7 * 52 * 4;
 
     address internal initializer;
     address public team;
     address public pendingTeam;
-    uint public teamRate;
-    uint public constant MAX_TEAM_RATE = 50; // 50 bps = 0.05%
+    uint256 public teamRate;
+    uint256 public constant MAX_TEAM_RATE = 50; // 5% max
```
Comment (0xMotto) [inline contracts/Minter.sol:31]: [Question] Just want to make sure we want 0.5% max team rate?

File: contracts/Minter.sol [hunk 12]
```diff
@@ -38,23 +44,23 @@ contract Minter is IMinter {
     ) {
         initializer = msg.sender;
         team = msg.sender;
-        teamRate = 30; // 30 bps = 0.03%
-        _velo = IVelo(IVotingEscrow(__ve).token());
+        teamRate = 30; // 30 bps = 3%
```
Comment (0xMotto) [inline contracts/Minter.sol:47]: [Question] Initial team rate is 0.3% here. Is this expected?

File: contracts/Minter.sol [hunk 13]
```diff
@@ -1,35 +1,41 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/Minter.sol:1]: [Note] mostly formatting changes in this contract

File: contracts/libraries/Math.sol [hunk 14]
```diff
@@ -1,16 +1,19 @@
 pragma solidity 0.8.13;
```
Comment (0xMotto) [inline contracts/libraries/Math.sol:1]: [Question] We can remove this contract as we are now using openzeppelin ones.

File: contracts/interfaces/IWrappedExternalBribeFactory.sol [hunk 15]
```diff
@@ -0,0 +1,5 @@
+pragma solidity 0.8.13;
```
Comment (0xMotto) [inline contracts/interfaces/IWrappedExternalBribeFactory.sol:1]: [Note] This interface will be used in `createGauge` function in `Voter` contract to create bribe.

File: contracts/interfaces/IBribe.sol [hunk 16]
```diff
@@ -1,9 +1,9 @@
 pragma solidity 0.8.13;
```
Comment (0xMotto) [inline contracts/interfaces/IBribe.sol:1]: [Note] Only formatting changes.

File: contracts/factories/WrappedExternalBribeFactory.sol [hunk 17]
```diff
@@ -4,14 +4,11 @@ pragma solidity 0.8.13;
 import {WrappedExternalBribe} from 'contracts/WrappedExternalBribe.sol';
 
 contract WrappedExternalBribeFactory {
-    address public immutable voter;
+    bool internal _initialized;
+    address public voter;
     mapping(address => address) public oldBribeToNew;
     address public last_bribe;
 
-    constructor(address _voter) {
```
Comment (0xMotto) [inline contracts/factories/WrappedExternalBribeFactory.sol:11]: [Question] why do we remove setting voter in constructor?

imo, if there is no specific reasons, please undo the changes in this contract

File: contracts/factories/WrappedExternalBribeFactory.sol [hunk 18]
```diff
@@ -4,14 +4,11 @@ pragma solidity 0.8.13;
...
-        voter = _voter;
-    }
-
     function createBribe(address existing_bribe) external returns (address) {
```
Comment (0xMotto) [inline contracts/factories/WrappedExternalBribeFactory.sol:12]: [Suggestion] If we confirm we need to remove voter in constructor, please:
1. add a require statement to make sure it is initialized and voter is set.
2. add in deployment script to setVoter -> already done here (https://github.com/Velocimeter/instruments/blob/tests/deploy/15_init.ts#L47)

File: contracts/factories/PairFactory.sol [hunk 19]
```diff
@@ -1,40 +1,96 @@
...
 contract PairFactory is IPairFactory {
-
     bool public isPaused;
     address public pauser;
     address public pendingPauser;
 
     uint256 public stableFee;
     uint256 public volatileFee;
-    uint256 public constant MAX_FEE = 5; // 0.05%
+    uint256 public constant MAX_FEE = 50; // 0.5%
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:14]: [Note] Max fee is changed from 0.05% to 0.5%

File: contracts/WrappedExternalBribe.sol [hunk 20]
```diff
@@ -1,46 +1,55 @@
 // SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/WrappedExternalBribe.sol:1]: [Note] only formatting changes.

File: contracts/InternalBribe.sol [hunk 21]
```diff
@@ -1,78 +1,87 @@
...
+import "contracts/interfaces/IVotingEscrow.sol";
 
 // Bribes pay out rewards for a given pool based on the votes that were received from the user (goes hand in hand with Voter.vote())
 contract InternalBribe is IBribe {
-
     address public immutable voter; // only voter can modify balances (since it only happens on vote())
     address public immutable _ve;
 
-    uint public constant DURATION = 7 days; // rewards are released over 7 days
-    uint public constant PRECISION = 10 ** 18;
```
Comment (0xMotto) [inline contracts/InternalBribe.sol:17]: [Note] Removing PRECISION and use token decimals.

File: contracts/Pair.sol [hunk 22]
```diff
@@ -1,129 +1,164 @@
...
-    event Approval(address indexed owner, address indexed spender, uint amount);
+    event Transfer(address indexed from, address indexed to, uint256 amount);
+    event Approval(address indexed owner, address indexed spender, uint256 amount);
 
     constructor() {
         factory = msg.sender;
-        (address _token0, address _token1, bool _stable) = PairFactory(msg.sender).getInitializable();
+        //voter = PairFactory(msg.sender).voter(); // nice easy way to add the voter :) we already getting this from pair factory tho
+        tank = PairFactory(msg.sender).tank(); // nice easy way to add the voter :) we already getting this from pair factory tho
+        (address _token0, address _token1, bool _stable) = PairFactory(msg.sender).getInitializable(); //wondering why msg.sender is passed here??
```
Comment (0xMotto) [inline contracts/Pair.sol:101]: msg.sender here is PairFactory because PairFactory creates Pair

File: contracts/RewardsDistributor.sol [hunk 23]
```diff
@@ -1,10 +1,10 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-License-Identifier: MIT
```
Comment (0xMotto) [inline contracts/RewardsDistributor.sol:1]: [Note] Just formatting change in this contract.

File: contracts/Router.sol [hunk 24]
```diff
@@ -2,15 +2,14 @@
 
 pragma solidity 0.8.13;
 
-import 'contracts/libraries/Math.sol';
-import 'contracts/interfaces/IERC20.sol';
-import 'contracts/interfaces/IPair.sol';
-import 'contracts/interfaces/IPairFactory.sol';
-import 'contracts/interfaces/IRouter.sol';
-import 'contracts/interfaces/IWETH.sol';
+import "openzeppelin-contracts/contracts/utils/math/Math.sol";
```
Comment (0xMotto) [inline contracts/Router.sol:5]: [Note] only formatting changes and this library in this contract.

File: contracts/Pair.sol [hunk 25]
```diff
@@ -1,129 +1,164 @@
...
 
-    event Transfer(address indexed from, address indexed to, uint amount);
-    event Approval(address indexed owner, address indexed spender, uint amount);
+    event Transfer(address indexed from, address indexed to, uint256 amount);
+    event Approval(address indexed owner, address indexed spender, uint256 amount);
 
     constructor() {
         factory = msg.sender;
-        (address _token0, address _token1, bool _stable) = PairFactory(msg.sender).getInitializable();
+        //voter = PairFactory(msg.sender).voter(); // nice easy way to add the voter :) we already getting this from pair factory tho
```
Comment (0xMotto) [inline contracts/Pair.sol:99]: [Suggestion] remove commented code please.

File: contracts/Pair.sol [hunk 26]
```diff
@@ -1,129 +1,164 @@
...
+        tank = PairFactory(msg.sender).tank(); // nice easy way to add the voter :) we already getting this from pair factory tho
+        (address _token0, address _token1, bool _stable) = PairFactory(msg.sender).getInitializable(); //wondering why msg.sender is passed here??
         (token0, token1, stable) = (_token0, _token1, _stable);
         fees = address(new PairFees(_token0, _token1));
+        //     externalBribe = address();  this does not need to be set at the time of creation
```
Comment (0xMotto) [inline contracts/Pair.sol:104]: [Suggestion] remove commented code please.

File: contracts/Pair.sol [hunk 27]
```diff
@@ -1,129 +1,164 @@
...
+    mapping(address => uint256) public nonces;
 
-    uint internal constant MINIMUM_LIQUIDITY = 10**3;
+    uint256 internal constant MINIMUM_LIQUIDITY = 10 ** 3;
 
     address public immutable token0;
     address public immutable token1;
     address public immutable fees;
-    address immutable factory;
+    address immutable factory; // explain this? can this be public? can I call this from inside a contract
```
Comment (0xMotto) [inline contracts/Pair.sol:37]: this is public

File: contractNameAndAddress.ts (hunk 2 above)
Comment (duncan4123) [reply contractNameAndAddress.ts:4]: @t0rbik

File: contracts/Minter.sol (hunk 11 above)
Comment (duncan4123) [reply contracts/Minter.sol:31]: yep

File: contracts/Minter.sol (hunk 12 above)
Comment (duncan4123) [reply contracts/Minter.sol:47]: yep

File: contracts/libraries/Math.sol (hunk 14 above)
Comment (duncan4123) [reply contracts/libraries/Math.sol:1]: does forge still use it?

File: contracts/factories/WrappedExternalBribeFactory.sol (hunk 17 above)
Comment (duncan4123) [reply contracts/factories/WrappedExternalBribeFactory.sol:11]: @t0rbik

File: contracts/factories/WrappedExternalBribeFactory.sol (hunk 18 above)
Comment (duncan4123) [reply contracts/factories/WrappedExternalBribeFactory.sol:12]: @t0rbik

File: contracts/Pair.sol (hunk 22 above)
Comment (duncan4123) [reply contracts/Pair.sol:101]: @t0rbik

File: contractNameAndAddress.ts (hunk 2 above)
Comment (t0rbik) [reply contractNameAndAddress.ts:4]: > @t0rbik

Script was introduced in [this commit](https://github.com/Velocimeter/instruments/pull/5/commits/8a2574389af638ffd8e8210a4d4fba831785e5cd).
It can be safely deleted as we don't use it, the type is obv wrong and doesn't exists in scope.

File: contracts/Gauge.sol (hunk 6 above)
Comment (t0rbik) [reply contracts/Gauge.sol:98]: Probably yes, based on information I've received today.
@duncan4123  could you elaborate on internal bribes contract managing fees as you've told today: 
> it will calculate it still using internal bribes

File: contracts/factories/WrappedExternalBribeFactory.sol (hunk 17 above)
Comment (t0rbik) [reply contracts/factories/WrappedExternalBribeFactory.sol:11]: > [Question] why do we remove setting voter in constructor?
> 
> imo, if there is no specific reasons, please undo the changes in this contract

There's a particular reason.
In order for voter to automate creation of wrapped external bribe, we added a wrapped external bribe factory to voter constructor. That's being said, we need to deploy wrapped external bribe factory before voter, thus voter can't be passed to constuctor. After contracts are deployed, idea is to setVoter in init script.

File: contracts/factories/WrappedExternalBribeFactory.sol (hunk 18 above)
Comment (t0rbik) [reply contracts/factories/WrappedExternalBribeFactory.sol:12]: > [Suggestion] If we confirm we need to remove voter in constructor, please:
> 
> 1. add a require statement to make sure it is initialized and voter is set.
> 2. add in deployment script to setVoter -> already done here (https://github.com/Velocimeter/instruments/blob/tests/deploy/15_init.ts#L47)

Could you be more specific about first point?

File: contracts/Pair.sol (hunk 22 above)
Comment (t0rbik) [reply contracts/Pair.sol:101]: I understand that is what we want @0xMotto @duncan4123

**Comment by t0rbik** [comment]
Went through comments, left some replies. I will go through again tomorrow and apply some suggestions.

File: contracts/libraries/Math.sol (hunk 14 above)
Comment (0xMotto) [reply contracts/libraries/Math.sol:1]: no test cases are using this.

File: contracts/factories/WrappedExternalBribeFactory.sol (hunk 18 above)
Comment (0xMotto) [reply contracts/factories/WrappedExternalBribeFactory.sol:12]: `createGauge` function of `Voter.sol` can be called before `voter` is set in `WrappedExternalBribeFactory`, so adding following require statement into `createBribe` function will help prevent it from happening.

```
require(_initialized, "Not initialized");
```

File: contracts/factories/WrappedExternalBribeFactory.sol [hunk 28]
```diff
@@ -21,4 +18,10 @@ contract WrappedExternalBribeFactory {
         oldBribeToNew[existing_bribe] = last_bribe;
         return last_bribe;
     }
+
+    function setVoter(address _voter) external {
```
Comment (0xMotto) [inline contracts/factories/WrappedExternalBribeFactory.sol:22]: please also add a requirement statement to restrict the caller.

```
require(msg.sender === team, "Not authorized");
```

File: contracts/factories/WrappedExternalBribeFactory.sol [hunk 29]
```diff
@@ -21,4 +18,10 @@ contract WrappedExternalBribeFactory {
...
+        require(!_initialized, "Already initialized");
```
Comment (0xMotto) [inline contracts/factories/WrappedExternalBribeFactory.sol:23]: what if we have made a mistake setting a wrong voter address?
current code will not allow further changing.

File: contractNameAndAddress.ts (hunk 2 above)
Comment (t0rbik) [reply contractNameAndAddress.ts:4]: I'm going to delete it.

File: contracts/factories/PairFactory.sol [hunk 30]
```diff
@@ -1,40 +1,96 @@
...
+        uint256
+    );
 
     constructor() {
         pauser = msg.sender;
         isPaused = false;
         feeManager = msg.sender;
-        stableFee = 2; // 0.02%
-        volatileFee = 2;
+        // stableFee = 2; // 0.02%
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:45]: Please remove commented code

File: contracts/Voter.sol [hunk 31]
```diff
@@ -215,22 +233,26 @@ contract Voter is IVoter {
...
-        address _gauge = IGaugeFactory(gaugefactory).createGauge(_pool, _internal_bribe, _external_bribe, _ve, isPair, allowedRewards);
+        address _wxbribe = IWrappedExternalBribeFactory(wrappedxbribefactory).createBribe(_external_bribe);
+        address _gauge = IGaugeFactory(gaugefactory).createGauge(
+            _pool, _internal_bribe, _external_bribe, _ve, isPair, allowedRewards
+        );
 
-        IERC20(base).approve(_gauge, type(uint).max);
+        IERC20(base).approve(_gauge, type(uint256).max);
         internal_bribes[_gauge] = _internal_bribe;
         external_bribes[_gauge] = _external_bribe;
```
Comment (0xMotto) [inline contracts/Voter.sol:257]: should change to following:

`external_bribes[_gauge] = _wxbribe`

File: contracts/Voter.sol [hunk 32]
```diff
@@ -215,22 +233,26 @@ contract Voter is IVoter {
...
             require(isPair, "!_pool");
             require(isWhitelisted[tokenA] && isWhitelisted[tokenB], "!whitelisted");
         }
 
         address _internal_bribe = IBribeFactory(bribefactory).createInternalBribe(internalRewards);
         address _external_bribe = IBribeFactory(bribefactory).createExternalBribe(allowedRewards);
-        address _gauge = IGaugeFactory(gaugefactory).createGauge(_pool, _internal_bribe, _external_bribe, _ve, isPair, allowedRewards);
+        address _wxbribe = IWrappedExternalBribeFactory(wrappedxbribefactory).createBribe(_external_bribe);
+        address _gauge = IGaugeFactory(gaugefactory).createGauge(
+            _pool, _internal_bribe, _external_bribe, _ve, isPair, allowedRewards
```
Comment (0xMotto) [inline contracts/Voter.sol:252]: please change `_external_bribe` to `_wxbribe `

File: contracts/Voter.sol [hunk 33]
```diff
@@ -1,88 +1,100 @@
...
     mapping(address => bool) public isWhitelisted;
     mapping(address => bool) public isAlive;
 
-    event GaugeCreated(address indexed gauge, address creator, address internal_bribe, address indexed external_bribe, address indexed pool);
+    event GaugeCreated(
+        address indexed gauge,
+        address creator,
+        address internal_bribe,
+        address indexed external_bribe,
+        address wxbribe,
```
Comment (0xMotto) [inline contracts/Voter.sol:52]: Suggest not to change this event but put wrapped external bribe contract address in `external_bribe`

File: contracts/Voter.sol [hunk 34]
```diff
@@ -239,7 +261,9 @@ contract Voter is IVoter {
         isAlive[_gauge] = true;
         _updateFor(_gauge);
         pools.push(_pool);
-        emit GaugeCreated(_gauge, msg.sender, _internal_bribe, _external_bribe, _pool);
+        Pair(_pool).setHasGauge(true); // may need to switch to IPair?
+        Pair(_pool).setExternalBribe(_wxbribe); // Changed this to wrapped external bribe  from
+        emit GaugeCreated(_gauge, msg.sender, _internal_bribe, _external_bribe, _wxbribe, _pool);
```
Comment (0xMotto) [inline contracts/Voter.sol:266]: https://github.com/Velocimeter/instruments/pull/5/files#r1110965391

Please amend according to the above comment

File: contracts/Voter.sol [hunk 35]
```diff
@@ -239,7 +261,9 @@ contract Voter is IVoter {
         isAlive[_gauge] = true;
         _updateFor(_gauge);
         pools.push(_pool);
-        emit GaugeCreated(_gauge, msg.sender, _internal_bribe, _external_bribe, _pool);
+        Pair(_pool).setHasGauge(true); // may need to switch to IPair?
```
Comment (0xMotto) [inline contracts/Voter.sol:264]: Add `setHasGauge` and `setExternalBribe` to `IPair` and use it.

File: contracts/Pair.sol [hunk 36]
```diff
@@ -1,129 +1,164 @@
...
 
-    function observationLength() external view returns (uint) {
+    function _safeApprove(address token, address spender, uint256 value) internal {
+        require(token.code.length > 0);
+        (bool success, bytes memory data) = token.call(abi.encodeWithSelector(IERC20.approve.selector, spender, value));
+        require(success && (data.length == 0 || abi.decode(data, (bool))));
+    }
+
+    function setExternalBribe(address _externalBribe) external {
+        // require(msg.sender == voter, "Only voter can set external bribe");
```
Comment (0xMotto) [inline contracts/Pair.sol:137]: Please use this require statement

File: contracts/Pair.sol [hunk 37]
```diff
@@ -1,129 +1,164 @@
...
+        externalBribe = _externalBribe;
+        _safeApprove(token0, externalBribe, type(uint256).max);
+        _safeApprove(token1, externalBribe, type(uint256).max);
+        // _safeApprove(token0, externalBribe, amount);
+    }
+
+    function setHasGauge(bool value) external {
+        // require(msg.sender == voter, "Only voter can set has gauge");
```
Comment (0xMotto) [inline contracts/Pair.sol:145]: Please use this require statement

File: contracts/Pair.sol (hunk 25 above)
Comment (0xMotto) [inline contracts/Pair.sol:99]: why don't we set voter here and use it to guard `setHasGauge` and `setExternalBribe`?

File: contracts/Pair.sol [hunk 38]
```diff
@@ -1,129 +1,164 @@
...
+        require(token.code.length > 0);
+        (bool success, bytes memory data) = token.call(abi.encodeWithSelector(IERC20.approve.selector, spender, value));
+        require(success && (data.length == 0 || abi.decode(data, (bool))));
+    }
+
+    function setExternalBribe(address _externalBribe) external {
+        // require(msg.sender == voter, "Only voter can set external bribe");
+        externalBribe = _externalBribe;
+        _safeApprove(token0, externalBribe, type(uint256).max);
+        _safeApprove(token1, externalBribe, type(uint256).max);
```
Comment (0xMotto) [inline contracts/Pair.sol:140]: Please do not give external bribe max approval...

File: contracts/Pair.sol [hunk 39]
```diff
@@ -1,129 +1,164 @@
...
+        // _safeApprove(token0, externalBribe, amount);
```
Comment (0xMotto) [inline contracts/Pair.sol:141]: Remove this.

File: contracts/Pair.sol [hunk 40]
```diff
@@ -148,45 +183,91 @@ contract Pair is IPair {
...
+    //     uint256 _ratio = (amount * 1e18) / totalSupply;
+    //     if (_ratio > 0) {
+    //         index1 += _ratio;
+    //     }
+    //     emit Fees(msg.sender, 0, amount);
+    // }
+
+    // Accrue fees on token0.
+    function _update0(uint256 amount) internal {
+        if (hasGauge == false) {
```
Comment (0xMotto) [inline contracts/Pair.sol:210]: do approval here

File: contracts/Pair.sol [hunk 41]
```diff
@@ -335,34 +426,42 @@ contract Pair is IPair {
...
-        // The curve, either x3y+y3x for stable pools, or x*y for volatile pools
-        require(_k(_balance0, _balance1) >= _k(_reserve0, _reserve1), 'K'); // Pair: K
+        uint256 amount0In = _balance0 > _reserve0 - amount0Out ? _balance0 - (_reserve0 - amount0Out) : 0;
+        uint256 amount1In = _balance1 > _reserve1 - amount1Out ? _balance1 - (_reserve1 - amount1Out) : 0;
+        require(amount0In > 0 || amount1In > 0, "IIA"); // Pair: INSUFFICIENT_INPUT_AMOUNT
+        {
+            // scope for reserve{0,1}Adjusted, avoids stack too deep errors
+            (address _token0, address _token1) = (token0, token1);
+            if (amount0In > 0) {
+                _update0((amount0In * PairFactory(factory).getFee(stable)) / 10000);
```
Comment (0xMotto) [inline contracts/Pair.sol:456]: Use `IPairFactory`

File: contracts/factories/PairFactory.sol [hunk 42]
```diff
@@ -1,40 +1,96 @@
...
+        // require(msg.sender == deployer); // have to make sure that this can be set to the voter addres during init script
+        voter = _voter;
+        initial_voter_set = true;
+    }
+
+    // function set tank on factory require team
+
+    function setTank(address _tank) external {
+        require(!initial_tank_set, "The tank has already been set.");
+        // require(msg.sender == deployer); // this should be updateable to team but adding deployer so that init script can run..
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:75]: Please use this require statement.

File: contracts/factories/PairFactory.sol [hunk 43]
```diff
@@ -1,40 +1,96 @@
...
+        team = _team;
+    }
+
+    // In this example, the owner variable is set to the address that deploys the contract in the constructor. The setTeam function requires that the caller (msg.sender) must be the owner, and that the team variable has not yet been set (it is equal to the address value of 0). If either of these conditions are not met, the function will revert and not update the team variable.
+
+    // we only get once shot at this.
+
+    function setVoter(address _voter) external {
+        require(!initial_voter_set, "The voter has already been set.");
+        // require(msg.sender == deployer); // have to make sure that this can be set to the voter addres during init script
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:66]: Please use this require statement.

File: contracts/factories/PairFactory.sol [hunk 44]
```diff
@@ -1,40 +1,96 @@
...
+    // This makes tank updateable forever by the team address (multisig)
+
+    function acceptTank(address _tank) external {
+        require(msg.sender == team, "not pending team");
+        tank = _tank;
+    }
+
+    // pair uses this to check if voter is updating external_bribe
+
+    function getVoter() external view returns (address) {
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:89]: we don't need this function.

Just call `PairFactory(address).voter()` in `Pair`

File: contracts/factories/PairFactory.sol [hunk 45]
```diff
@@ -54,46 +110,60 @@ contract PairFactory is IPairFactory {
     }
 
     function setFeeManager(address _feeManager) external {
-        require(msg.sender == feeManager, 'not fee manager');
+        require(msg.sender == feeManager, "not fee manager");
```
Comment (0xMotto) [inline contracts/factories/PairFactory.sol:113]: what happen if deployer key is compromised

File: contracts/Pair.sol [hunk 46]
```diff
@@ -148,45 +183,91 @@ contract Pair is IPair {
...
 
+    // dunks update0
+
     // Accrue fees on token0
-    function _update0(uint amount) internal {
-        _safeTransfer(token0, fees, amount); // transfer the fees out to PairFees
-        uint256 _ratio = amount * 1e18 / totalSupply; // 1e18 adjustment is removed during claim
-        if (_ratio > 0) {
-            index0 += _ratio;
+    // function _update0(uint256 amount) internal {
```
Comment (0xMotto) [inline contracts/Pair.sol:189]: remove commented code

File: contracts/Pair.sol [hunk 47]
```diff
@@ -148,45 +183,91 @@ contract Pair is IPair {
...
     }
 
     // Accrue fees on token1
-    function _update1(uint amount) internal {
-        _safeTransfer(token1, fees, amount);
-        uint256 _ratio = amount * 1e18 / totalSupply;
-        if (_ratio > 0) {
-            index1 += _ratio;
+    function _update1(uint256 amount) internal {
+        if (hasGauge == false) {
```
Comment (0xMotto) [inline contracts/Pair.sol:232]: do approval here

File: contracts/Voter.sol (hunk 32 above)
Comment (duncan4123) [reply contracts/Voter.sol:252]: Doesn't the factory want the address of the external bribe contract though?

Comment (duncan4123) [reply contracts/Voter.sol:252]: Why would we change this?

File: contracts/Voter.sol (hunk 31 above)
Comment (duncan4123) [reply contracts/Voter.sol:257]: I don't know why we would change this. The wrapped bribe in velodrome functions correctly without this change.

File: contracts/VeArtProxy.sol [hunk 48]
```diff
@@ -33,7 +33,7 @@ contract VeArtProxy is IVeArtProxy {
         output = string(abi.encodePacked(output, "locked_end ", toString(_locked_end), '</text><text x="10" y="80" class="base">'));
         output = string(abi.encodePacked(output, "value ", toString(_value), '</text></svg>'));
 
-        string memory json = Base64.encode(bytes(string(abi.encodePacked('{"name": "lock #', toString(_tokenId), '", "description": "Velodrome locks, can be used to boost gauge yields, vote on token emission, and receive bribes", "image": "data:image/svg+xml;base64,', Base64.encode(bytes(output)), '"}'))));
+        string memory json = Base64.encode(bytes(string(abi.encodePacked('{"name": "lock #', toString(_tokenId), '", "description": "Velocimeter locks, can be used to boost gauge yields, vote on token emission, and receive bribes", "image": "data:image/svg+xml;base64,', Base64.encode(bytes(output)), '"}'))));
```
Comment (duncan4123) [inline contracts/VeArtProxy.sol:36]: Would be good to do some work on this.

File: contracts/Voter.sol (hunk 32 above)
Comment (0xMotto) [reply contracts/Voter.sol:252]: i just saw in the test cases that external bribe address will be checked from GaugeFactory and used to call `notifyReward` . Not sure if frontend / api do the same
